apis/v1alpha2/httproute_types.go

```diff
@@ -0,0 +1,752 @@
+	// This can change in the future based on feedback during the alpha stage.
+	//
+	// Conformance-levels at this level are defined based on the type of filter:
+	//
+	// - ALL core filters MUST be supported by all implementations.
+	// - Implementers are encouraged to support extended filters.
+	// - Implementation-specific custom filters have no API guarantees across
+	//   implementations.
+	//
+	// Specifying a core filter multiple times has unspecified or custom conformance.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Filters []HTTPRouteFilter `json:"filters,omitempty"`
+
+	// BackendRefs defines the backend(s) where matching requests should be
+	// sent. If unspecified or invalid (refers to a non-existent resource or a Service with no endpoints),
+	// the rule performs no forwarding; if no filters are specified that would result in a
+	// response being sent, a HTTP 503 status code is returned. 503 responses must be sent so that the overall
+	// weight is respected; if an invalid backend is requested to have 80% of requests, then 80% of requests
+	// must get a 503 instead.
+	//
+	// Support: Core for Kubernetes Service
+	// Support: Custom for any other resource
+	//
+	// Support for weight: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	BackendRefs []HTTPBackendRef `json:"backendRefs,omitempty"`
+}
+
+// PathMatchType specifies the semantics of how HTTP paths should be compared.
+// Valid PathMatchType values are:
+//
+// * "Exact"
+// * "Prefix"
+// * "RegularExpression"
+// * "ImplementationSpecific"
+//
+// Prefix and Exact paths must be syntactically valid:
+//
+// - Must begin with the '/' character
+// - Must not contain consecutive '/' characters (e.g. /foo///, //).
+// - For prefix paths, a trailing '/' character in the Path is ignored,
+// e.g. /abc and /abc/ specify the same match.
+//
+// +kubebuilder:validation:Enum=Exact;Prefix;RegularExpression;ImplementationSpecific
+type PathMatchType string
+
+// PathMatchType constants.
+const (
+	PathMatchExact                  PathMatchType = "Exact"
+	PathMatchPrefix                 PathMatchType = "Prefix"
+	PathMatchRegularExpression      PathMatchType = "RegularExpression"
+	PathMatchImplementationSpecific PathMatchType = "ImplementationSpecific"
+)
+
+// HTTPPathMatch describes how to select a HTTP route by matching the HTTP request path.
+type HTTPPathMatch struct {
+	// Type specifies how to match against the path Value.
+	//
+	// Support: Core (Exact, Prefix)
+	//
+	// Support: Custom (RegularExpression, ImplementationSpecific)
+	//
+	// Since RegularExpression PathType has custom conformance, implementations
+	// can support POSIX, PCRE or any other dialects of regular expressions.
+	// Please read the implementation's documentation to determine the supported
+	// dialect.
+	//
+	// +optional
+	// +kubebuilder:default=Prefix
+	Type *PathMatchType `json:"type,omitempty"`
```

Comment:
this is not validated.

Reply:
This is a bit of a strange trick we're using throughout the API where we add the validation to a generic type like this so we can reuse that validation without redefining it anywhere it's used. In this case the validation is hiding here: https://github.com/kubernetes-sigs/gateway-api/blob/master/apis/v1alpha2/httproute_types.go#L233-L234